src/envoy/alts/tsi_transport_socket.cc

```diff
@@ -0,0 +1,278 @@
+
+  if (status != TSI_INCOMPLETE_DATA && status != TSI_OK) {
+    ENVOY_CONN_LOG(debug, "TSI: Handshake failed: status: {}",
+                   callbacks_->connection(), status);
+    return Network::PostIoAction::Close;
+  }
+
+  if (next_result->to_send_->length() > 0) {
+    raw_write_buffer_.move(*next_result->to_send_);
+  }
+
+  if (status == TSI_OK && handshaker_result != nullptr) {
+    tsi_peer peer;
+    tsi_handshaker_result_extract_peer(handshaker_result, &peer);
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: peer properties: {}",
+                   callbacks_->connection(), peer.property_count);
+    for (size_t i = 0; i < peer.property_count; ++i) {
+      ENVOY_CONN_LOG(debug, "  {}: {}", callbacks_->connection(),
+                     peer.properties[i].name,
+                     std::string(peer.properties[i].value.data,
+                                 peer.properties[i].value.length));
+    }
+    tsi_peer_destruct(&peer);
+
+    const unsigned char *unused_bytes;
+    size_t unused_byte_size;
+
+    status = tsi_handshaker_result_get_unused_bytes(
+        handshaker_result, &unused_bytes, &unused_byte_size);
+    ASSERT(status == TSI_OK);
+    if (unused_byte_size > 0) {
+      raw_read_buffer_.add(unused_bytes, unused_byte_size);
+    }
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: unused_bytes: {}",
+                   callbacks_->connection(), unused_byte_size);
+
+    status = tsi_handshaker_result_create_frame_protector(
+        handshaker_result, &max_output_protected_frame_size_,
+        &frame_protector_);
+    ASSERT(status == TSI_OK);
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: max frame: {}",
+                   callbacks_->connection(), max_output_protected_frame_size_);
+
+    tsi_handshaker_result_destroy(handshaker_result);
+
+    handshake_complete_ = true;
+    callbacks_->raiseEvent(Network::ConnectionEvent::Connected);
+  }
+
+  if (raw_read_buffer_.length() > 0) {
+    callbacks_->setReadBufferReady();
+  }
+  return Network::PostIoAction::KeepOpen;
+}
+
+Network::IoResult TsiSocket::doRead(Buffer::Instance &buffer) {
+  Network::IoResult result = raw_buffer_socket_.doRead(raw_read_buffer_);
+  ENVOY_CONN_LOG(debug, "TSI: raw_read result action {} bytes {} end_stream {}",
+                 callbacks_->connection(), enumToInt(result.action_),
+                 result.bytes_processed_, result.end_stream_read_);
+  if (result.action_ == Network::PostIoAction::Close &&
+      result.bytes_processed_ == 0) {
+    return result;
+  }
+
+  if (!handshake_complete_) {
+    Network::PostIoAction action = doHandshake();
+    if (action == Network::PostIoAction::Close || !handshake_complete_) {
+      return {action, 0, false};
+    }
+  }
+
+  if (handshake_complete_) {
+    size_t message_size = raw_read_buffer_.length();
+    auto *message_bytes = reinterpret_cast<unsigned char *>(
+        raw_read_buffer_.linearize(message_size));
+
+    unsigned char unprotected_buffer[4096];
```

Comment:
Is there any reason that we set buffer size to 4096?

Reply:
No particular, it's from sample code of TSI, I will tune later, added a comment